File: 1.Elementary level/9.Additional-Concept/Utils/Protocols/1.Please_Read./example/Circuit/Circuit.py
import datetime, time
class Circuit:
    def __init__(self, node1,node2):
        self.node1 = node1
        self.node2 = node2
        self.path = None
        self.bandwidth = { "one_second": "1k" }
        self.resources = []
        self.communications = None
        self.startTime = None
    def connect(self):
        if self.path != None:
            print("Circuit is already established.")
            return
        self.startTime = datetime.datetime.now()
        self.path = {"time":self.startTime,"switches":[self.node1,self.node2]}  
        self.resources.append(self.bandwidth)
        self.communications = {
            "connections":{
                "one": self.node1,
                "two": self.node2,
                "bandwidth": self.bandwidth,
                "resources": self.resources,
                "path": self.path,
                "time": self.startTime,
            }
        }
        print("Circuit established between",self.node1,"and",self.node2)
    def disconnect(self) :
        if self.path  == None:
          return print("No active circuit to disconnect.")
        endTime =  datetime.datetime.now()
        duration = (endTime - self.startTime) / 1000

        print(f"Circuit disconnected. Duration: {duration} seconds")
        self.path = None
        self.resources = []
        self.communications = None
        self.startTime = None

class Node:
    def __init__(self,name):
        self.name = name
        self.resources = []

    def connect(self):
     print(self.name, "is connecting...")
    

    def disconnect(self):
        self.resources = []
        print(self.name, "is disconnecting...")

if __name__ == "__main__":
    nodeA = Node("Node A")
    nodeB = Node("Node B")
    circuit = Circuit(nodeA, nodeB)

    circuit.connect()
    time.sleep(5)
    circuit.disconnect()

File: 1.Elementary level/9.Additional-Concept/Utils/Protocols/1.Please_Read./example/Circuit/Circuit.js
class Circuit {
  constructor(node1, node2) {
    this.node1 = node1;
    this.node2 = node2;
    this.path = null;
    this.bandwidth = { one_second: "1k" };
    this.resources = [];
    this.communications = null;
    this.startTime = null;
  }

  connect() {
    if (this.path !== null) {
      console.log("Circuit is already established.");
      return;
    }

    this.startTime = new Date();
    this.path = { time: this.startTime, switches: [this.node1, this.node2] };
    this.resources.push(this.bandwidth);
    this.communications = {
      connections: {
        one: this.node1,
        two: this.node2,
        bandwidth: this.bandwidth,
        resources: this.resources,
        path: this.path,
        time: this.startTime,
      },
    };

    console.log("Circuit established between", this.node1, "and", this.node2);
  }

  disconnect() {
    if (this.path === null) {
      console.log("No active circuit to disconnect.");
      return;
    }

    const endTime = new Date();
    const duration = (endTime - this.startTime) / 1000;

    console.log(`Circuit disconnected. Duration: ${duration} seconds`);

    this.path = null;
    this.resources = [];
    this.communications = null;
    this.startTime = null;
  }
}

class Node {
  constructor(name) {
    this.name = name;
    this.resources = [];
  }

  connect() {
    console.log(this.name, "is connecting...");
  }

  disconnect() {
    this.resources = [];
    console.log(this.name, "is disconnecting...");
  }
}
const nodeA = new Node("Node A");
const nodeB = new Node("Node B");
const circuit = new Circuit(nodeA, nodeB);

circuit.connect();
setTimeout(() => {
  circuit.disconnect();
}, 5000);
